Conditions 599 Refactor ComboBox

williamphelps13 opened the pull request:
### Merged Prerequisite - Merge this vets-api PR to create the feature toggle - [Conditions 599 Add "Add Disabilities Enhancement" feature toggle #18983](https://github.com/department-of-veterans-affairs/vets-api/pull/18983)

## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to TeamSites and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary
- The goal for this PR when deployed to prod would be to NOT change the user experience. 
  - This PR has a feature toggle that when disabled (which it will be in staging and prod) will have the Add Disabilities page use all the previous code including the ComboBox. 
  - After this PR is merged and we ensure the it is working like it was before in staging, we can enable the feature toggle in staging. 
  - This will then allow us to ensure that staging is working as we would like prod to work. 
  - When we are satisfied we can enable the feature toggle in staging. If all works correctly we can remove this feature toggle shortly after.
- [Form526EZApp.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-7160eda0f93aa9a0dc5391889d45cf5fd5ab5c33ecd281c9c163aa2f5bc452bd) 
- [Autocomplete.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-3db3743138b9557bbf5cd29951c87931037da99d5be0d55b7f06d681c41eb780)
  - Updates name of component to Autocomplete ([from advice from Matt Dingee](https://dsva.slack.com/archives/C8E985R32/p1729808871239759))
  - Updates component to React Functional component
  - Reduces lines of code from 450 to 185
  - Simplifies state by updating from 2 states (searchTerm and value) to 1 state (value) for tracking the input value
  - Simplifies aria-live to no longer require 2 html elements and 3 pieces of state (bump, ariaLive1, and ariaLive2) and instead requiring just 1 element and 1 piece of state (ariaLiveText)
  - Makes debounceTime that was aimed to be passed and used by the component actually passed and used
  - Highlights freeText result when typing solving [Highlight free text option #512](https://app.zenhub.com/workspaces/contention-classification-640a24db5dcd08457cebbc79/issues/gh/department-of-veterans-affairs/vagov-claim-classification/512)
  - Removes bolding within freeText result
    - Makes the code implementation much cleaner while reducing queuing of user to select free text
    - We actually would prefer if the user selects a search result rather then the free text result. The `Enter your condition as "${val}"` without the bold seems to clearly tell the user that if they select that option they are entering the value within the quotes. Therefore, the bold may increase the likelihood of selecting the free text because of how bold text attracts attention (an undesired outcome) while not providing much additional clarity to the user.
  - Keeps focus in the input as arrow through results which upon [asking platform accessibility](https://dsva.slack.com/archives/C8E985R32/p1728398775821319) had consensus as a better user experience because the user was visually queued that if they type the text will go into the input. Also matches with the following accessible  examples:
    - [Downshift](https://www.downshift-js.com/downshift/) (demos are on the link page underneath the code blocks) - keeps the focus in the input as the user arrows down through results (the highlighting effect is not focus)
    - [ARIA APG Editable Combobox With List Autocomplete](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/) (example is in the Example section)
  - Fixes issues that when user blurred input, input was automatically refocused
    - Aligns functionality that the first form input is focused upon hitting tab but not on page load 
    - This fixes confusing user experience as it not a common experience for an input to refuse to be unfocused
    - Fixes test case of "If click 'Add another condition' while another condition is in edit mode, focus should jump to the new input field" - previously focus would not leave first input field
  - Fixes issue that when user adds value then blurs input then re-focuses input, the list would not be shown until the user changed the value. Now it opens the list on re-focus. 
    - This aligns with common-sense UX
    - It also aligns with the behavior of [Downshift](https://www.downshift-js.com/downshift/) and [ARIA APG Editable Combobox With List Autocomplete](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/).
  - Fixes continued arrow down in list when at bottom of viewable part of list resulted in a jolting and confusing experience. Now arrow down works predictably. 
    - Replaced `behavior: smooth` with `behavior: auto` so that on holding down of arrow there is not longer a confusing delay - `behavior: auto` is the default so I then deleted this property 
  - Fixed "index.js:287 New axe issues index.js:307 critical: ARIA attributes must conform to valid values" on type into field  
- [constants.js](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-baae3f550e35220e2d29206d29c8649813021f096b61876c98a6bb9fbe354b1e)
- [addDisabilities.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-4631fe5289b967d896025bdd9b6777a0845b7b1c2b47801e70a04c464553c715)
- [addDisabilities.js](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-7a9dd7c424ab9980579d6167e1db4743e1b1f20b532d65b00e9eaaee2c20a75f)
- [disability-benefits.scss](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-2dfb28e808a1a93e8164cdb92b923ef5ea103d81114c84c60269548152db9d9f)
- [Autocomplete.unit.spec.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-253a365b44dbb8432f96c45482d8a26646b3f8194281342d346ddb6c2d4d4948)
- [cypress.helpers.js](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-e884ea03b47a9baa004b871fd4f054b02ada591cbcc9e0f5435d590da463a46d)
- [featureFlagNames.json](https://github.com/department-of-veterans-affairs/vets-website/pull/32311/files#diff-dd0d0ea85b2741bb70a428d6d62788694aedf62e624eb758f65800d421147082)

- Which team do you work for, does your team own the maintenance of this component? Conditions Team and Yes


## Related issue(s)

- [Refactor ComboBox into modular React functional component #599](https://app.zenhub.com/workspaces/contention-classification-640a24db5dcd08457cebbc79/issues/gh/department-of-veterans-affairs/vagov-claim-classification/599)
- [Highlight free text option #512](https://app.zenhub.com/workspaces/contention-classification-640a24db5dcd08457cebbc79/issues/gh/department-of-veterans-affairs/vagov-claim-classification/512)

## Testing done
- _Describe what the old behavior was prior to the change_
- Describe the steps required to verify your changes are working as expected:
  - [x] UI matches final designs
    - <img width="204" alt="Screenshot 2024-10-10 at 9 28 27 AM" src="https://github.com/user-attachments/assets/6d960af8-ef00-4cbf-a4c6-dd5bbde963ba">
    - <img width="209" alt="Screenshot 2024-10-10 at 9 28 06 AM" src="https://github.com/user-attachments/assets/10ad3286-14b1-4cb1-a07a-4cb0849382f6">
  - [ ] When load the page, the first interactive element in the conditions page section should be highlighted and when click tab the focus moves to the next interactive element on the page, etc. and so on, in the order expected.
  - [x] When adding a new condition and the condition text is empty and click save, error text should display that user cannot save empty condition
  - [x] When the user is typing, the String Matching algorithm should return the free text result and then a max of 20 search results 
  - [ ] When the user is typing, the dropdown max height will be set to display only 10 terms at a time
  - [x] When the user is typing, the free-text element should be highlighted and if the user scrolls or navigates with the arrows keys, the selection should move to the next elements
  - [ ] When the user is typing, the free text input is the first element in the list, the option will highlight the matching text in bold, and there should be a line below to distinguish this free-text option 
  - [x] When the user browses the dropdown list of elements, the Screen Reader should announce the active option
  - [ ] Headings, aria labels and labels/for align with CAIA feedback
  - [x] When adding new conditions, new conditions should be added to the list in "Read Mode"
  - [x] "Add another condition" button should be clickable
  - [x] After clicking save or update, focus should move to the edit button of the added/updated condition
  - [x] The user tries to save the same condition as an already existing condition, an error message should display that this condition was already added and the condition should remain in Edit mode
  - [x] When user saves a condition, capitalize the first letter of the condition
  - [x] If click "Add another condition" while another condition is in edit mode, focus should jump to the new input field
  - [x] When click edit, text-input will populate with condition and focus moves to end of the condition in the input field
  - [x] When editing condition and the condition text is emptied and then click update, error text should display that cannot save empty condition
  - [x] The user added more than one condition and one of the  condition is in "Edit Mode" and click "Remove", an error message appears when user tries to remove without having any conditions. "You must enter at least one condition to remove". 
  - [x] When editing condition, the new value in the input should replace the previous saved value
  - [x] When editing condition and click "Cancel", condition should return to "Read Mode" with the previous value it had
  - [x] When editing condition and click "Remove", condition should be removed from the list
 
- Describe the tests completed and the results:
  - THIS AND PREVIOUS VERSION DID NOT FOCUS ANY ELEMENT BY DEFAULT: When load the page, the first interactive element in the conditions page section should be highlighted and when click tab the focus moves to the next interactive element on the page, etc. and so on, in the order expected.
  - CHANGED:  When click edit, dropdown not visible until user types --> Dropdown is visible so that they can select a different option without needing to delete a character to see the options
  - CHANGED: When the user is typing, the free-text element should be highlighted and if the user scrolls or navigates with the arrows keys, the selection should move to the next elements --> Changed smooth scroll to auto scroll for holding arrow down functionality
  - CHANGED:  When the user is typing, The dropdown max height will be set to display only 10 terms at a time --> changed this to 9 terms so that the dropdown does not overlap the footer and blur with that content
  - CHANGED: When the user is typing, the option will highlight the matching text in bold --> This is definitely doable but could not find a clean and readable way to implement AND also worried that bold may be encouraging the user to select the option. Are the quotes not enough to communicate to the user that that text is the same as their input value?
  - NOTE: When editing condition and click "Cancel", condition should return to "Read Mode" with the previous value it had - Cancel and update only appearing on first array item

## Screenshots  staging on left, PR code on right)
<img width="1377" alt="Screenshot 2024-10-21 at 9 00 59 AM" src="https://github.com/user-attachments/assets/584a660d-26c8-48dc-9880-400411172993">
<img width="1384" alt="Screenshot 2024-10-21 at 8 58 35 AM" src="https://github.com/user-attachments/assets/2c9560b6-df38-4da1-afca-4291d426e650">
<img width="1373" alt="Screenshot 2024-10-21 at 9 00 06 AM" src="https://github.com/user-attachments/assets/e71e4ff0-b09d-4ad7-ad43-02b348eb9e71">

## Acceptance criteria

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [x] Events are being sent to the appropriate logging solution
- [x] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Requested Feedback

Is there a better feature toggle implementation?

williamphelps13 committed:
Refactor ComboBox (rename to AutoComplete) and related content

williamphelps13 committed:
Update tests to all be passing with updated functional component

williamphelps13 committed:
Update e2e AutoComplete selector

williamphelps13 committed:
Merge branch 'main' into conditions-599-refactor-combobox

bacitracin commented on src/applications/disability-benefits/all-claims/components/AutoComplete.jsx:
```diff
@@ -0,0 +1,191 @@
+                </li>
+              ))}
+            </ul>
+            <p role="alert" className="vads-u-visibility--screen-reader">
```
We'll want to add back in `aria-live="polite" aria-atomic="true"` here and below. Otherwise every time you land on a dropdown result you get an "alert" sound.

bacitracin commented:
I've been seeing some funkiness with the freetext appearing twice. I'm not totally clear on how to reproduce, but I've had the best luck by editing a condition and removing a couple letters
<img width="951" alt="Screenshot 2024-10-09 at 9 52 00 AM" src="https://github.com/user-attachments/assets/d46f88dc-cfb4-4164-a854-1b48a79f44ba">

williamphelps13 committed:
Refine AutoComplete component including PR comment fixes

Remove delayAnnouncement, fix eslint error by switching to useRef, make freeTextResult be full displayed text again, remove isOpen, refactor navigateList logic, add input keeps focus on selection functionality, and update to aria-live polite / atomic. Fix test from freeText and aria-live change.

bacitracin commented:
Can we add back in 
- the "{{condition}} is selected" announcement on selection
- and the "Input is empty. please enter a condition" announcement when the user has an input and then deletes all their text? 

I saw you added that text to the greater autocomplete instructions and I think we can remove it from that, because those autocomplete instructions are announced together with the label "Enter a condition" - it's duplicated a bit.

williamphelps13 committed:
Merge branch 'main' into conditions-599-refactor-combobox

bacitracin commented on src/applications/disability-benefits/all-claims/components/AutoComplete.jsx:
```diff
@@ -0,0 +1,187 @@
+import React, { useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+import { VaTextInput } from '@department-of-veterans-affairs/component-library/dist/react-bindings';
+import debounce from 'lodash/debounce';
```
Not sure if this is due to our debounce setting or what, but in VO + Safari, if I typed quickly, I was able to get way behind and mismatch the voiceover announcements by several steps. Eventually things would catch up, but by then a different item might be selected than what was announcing. 

In this case I just deleted my text and it never updated me with the input text changing at all <img width="850" alt="Screenshot 2024-10-10 at 10 30 01 AM" src="https://github.com/user-attachments/assets/38eb8e6d-de60-4ba7-a98c-48de90381915">

bacitracin replied:
I don't know how reproducable this is, so let's just keep an eye on it / be sure to test.

williamphelps13 committed:
Git merge main

williamphelps13 committed:
Attempt to add selected and empty aria message and to simplify ariaLive handling using just 1 element

williamphelps13 committed:
Move aria-live outside of list conditional so that it is still read for selected and empty

williamphelps13 committed:
Update e2e test selector and revert list height back to 440px

rmessina1010 reviewed (dismissed):
LGTM

williamphelps13 committed:
Add debouncedSetAriaLiveText, handleKeyDown list conditional, and remove aria-atomic

Add debouncedSetAriaLiveText so as to delay screen reader result count reading to avoid interruption by input content reading, only fire handleKeyDown when there is a list, and remove aria-atomic because there is no parent node of the aria-live we need it to read.

williamphelps13 committed:
Restore ComboBox and add feature toggle that only when enabled uses autocomplete

williamphelps13 committed:
Add id to AutoComplete, revert ComboBox to have its own data-testid, and fix scss duplication

williamphelps13 committed:
Temporarily remove addDisabilitiesAutocomplete testing to show passing

williamphelps13 committed:
Remove autcomplete definition and wrapper, update to debounceDelay, simplify conditionals

williamphelps13 committed:
Organize element attributes and extend blur timeout to ensure working click

williamphelps13 committed:
Merge branch 'main' into conditions-599-refactor-combobox

bacitracin commented on src/applications/disability-benefits/all-claims/components/AutoComplete.jsx:
```diff
@@ -0,0 +1,177 @@
+    debouncedSearch(inputValue);
+  };
+
+  const scrollIntoView = index => {
```
Just FYI no action needed: On my other team I'm doing some work with the scrolling behavior of the app not honoring `prefers-reduced-motion`, so when that ticket is done we should remember to come back and check what this combobox experience is like for someone who has that accessibility setting turned on https://github.com/department-of-veterans-affairs/va.gov-team/issues/95517

bacitracin commented on src/applications/disability-benefits/all-claims/components/AutoComplete.jsx:
```diff
@@ -0,0 +1,177 @@
+        ref={inputRef}
+        required
+        value={value}
+        onBlur={() => setTimeout(closeList, 200)} // Enables clicking option from list
```
Best practice is usually to pull inline functions out since they create a new function instance on every render. I don't think we'd really get that much performance benefit, and it does create more functions that aren't reused, so up to you.

bacitracin commented on src/applications/disability-benefits/all-claims/components/AutoComplete.jsx:
```diff
@@ -0,0 +1,177 @@
+      />
+      {results.length > 0 && (
+        <ul
+          aria-hidden="true"
```
I don't think that we'll want to aria hide the listbox all the time, just when it's not visible. Sighted screenreader users will be confused why they can't navigate to the ul and check out the contents before making a selection. Also having the aria-hidden here may mess with whether or not the component is shown on the forms shortcut pages

bacitracin replied:
Just noting that Will and I paired on this and came to the decision that this works. It feels so wrong to straight up put an aria-hidden true on an element we're interacting with , but the focus is technically staying in the input.

williamphelps13 committed:
Move inline function to handleBlur

williamphelps13 committed:
Update in files from AutoComplete to Autocomplete

williamphelps13 committed:
Change file name from AutoComplete to Temp (next commit updates to Autocomplete)

williamphelps13 committed:
Change file name from Temp to Autocomplete

tyler-spangler6 reviewed (approved):
LGTM. I tested the feature toggle in localhost and it hid the changes as expected. After toggling the feature flag on/off and logging out and then logging back in the correct component was shown based on the feature toggle.

williamphelps13 committed:
Merge branch 'main' into conditions-599-refactor-combobox

williamphelps13 committed:
Merge branch 'main' into conditions-599-refactor-combobox
